### PR TITLE
Fix #876: Allow implicit conversions from singleton types

### LINKED ISSUE #876: Allow implicit conversions to have a parameter of singleton type?
This is explicitly disallowed since https://github.com/lampepfl/dotty/commit/f0d34873c27cfbf7dab3909b9855dfeae4e8506e the commit message says:

> Reason for 2) is that we can do a much better caching for eligible implicits if we can widen singleton arguments in ViewProto. This leads to consicderable speedups in implicit search which seem to be hard to get without the restriction.

But implicit conversions on singletons seem to be a common pattern, see for example:
- https://github.com/molecule-labs/molecule/blob/master/molecule-core/src/main/scala/molecule/stream/IChan.scala#L807
- https://github.com/scalameta/scalameta/blob/master/scalahost/src/main/scala/scala/meta/internal/hosts/scalac/converters/ToMname.scala#L36

We should either allow them or find a good replacement for them that can be automatized.

### PR DESCRIPTION
**Pull request by odersky**
I found a way to make this efficient (hopefully).

### CODE REVIEW

**Comment by allanrenucci** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/typer/Implicits.scala [hunk 1]
```diff
@@ -71,17 +71,20 @@ object Implicits {
     /** The implicit references */
     def refs: List[ImplicitRef]
 
+    private var SingletonClass: ClassSymbol = null
```
Comment (allanrenucci) [inline compiler/src/dotty/tools/dotc/typer/Implicits.scala]: `private[this]` to avoid generating accessors?

**Comment by odersky** [comment]
The benchmark numbers look good after tweak. So this is ready to go in, from my side.

File: compiler/src/dotty/tools/dotc/typer/Implicits.scala [hunk 2]
```diff
@@ -71,6 +71,15 @@ object Implicits {
...
```
[the remark this answers is not in the document]
Comment (odersky) [reply compiler/src/dotty/tools/dotc/typer/Implicits.scala]: Same reason as for the caches in TypeComparer. It's used a lot and as a lazy val access is not super fast. It was one of the performance tweaks I applied after we saw a slip. Not sure it was the one that made the difference.

**Pull request merged**